docs/design/resolve-divergence.md

```diff
@@ -0,0 +1,104 @@
+
+### GetCommitDescription
+
+We will provide an optional command-line `--description-source <COMMIT_ID>` argument for specifying which commit-id to use as the source of the description. If specified this must be one of the commits in $DivergentCommits$. If `--description-source` is not specified and all commits in $DivergentCommits$ have identical description, we use that. Otherwise we bail out with a message asking the user to rerun the command with `--description-source <COMMIT_ID>`. We should probably also allow `-m <DESCRIPTION>` as an alternative solution.
+
+### $ResolveDivergentChange(C, DivergentCommits, EvologGraph, Parents, Description, mutable Repo)$
+
+ResolveDivergentChange operates EvologGraph, modifying it in place (in-memory). The steps of the main loop in ResolveDivergentChange are:
+
+1. Starting at the root $R$ of $EvologGraph$.
+1. If $R$ has no successors we exit this loop.
+1. Put the successors of $R$ in one of two sets: those that have no predecessor other than $R$ are added to set $next$, the rest are added to $remaining$.
+1. MergeDivergentTrees: merges the trees of the commits in $next$ with $R$ as base, producing $R'$ (see below). This may result in merge conflicts.
+1. If $R$ is in $DivergentCommits$, add its commit-id to $CommitsToHide$ and remove it from $DivergentCommits$.
+1. Add the intersection of $next$ and $DivergentCommits$ to $CommitsToHide$ and remove them from $DivergentCommits$.
+1. Add $R'$ to $DivergentCommits$.
+1. For each comming $X$ in $remaining$: replace the $R->X$ edge and any edge from $next$ to $X$ with an edge from $R'$ to $X$, while keeping other predecessors of $X$. In other words, $X$ is now a successor of $R'$.
+1. Similarly, for each successor $X$ of any commit in $next$, make $X$ a successor of $R'$, while keeping other predecessors of $X$ (if any).
+1. Set $R$ = $R'$ and go back to step 2.
+
+We repeat the steps above until we exit the loop with a single commit $R$ with the given description and parents. In the happy case the tree in commit $R$ does not have conflicts, but sometimes it will have conflicts. ResolveDivergentChange returns a summary of the changes made, including $R$ and $CommitsToHide$.
+
+The algorithm describe above moves in the "forward" direction. A similar "backwards" algorithm (possibly simpler?) can be implemented. It is hard to say if one is better than the other, it probably depends heavily on the specific situation. It is quite possible we will need heuristics to decide when to give up, for example if conflicts with many sides are produced.
+
+### MergeDivergentTrees
+
+This takes $successors$ (a non-empty set of commits), and $base$ (a single commit). If $successors$ has just one commit, returns that commit. Say $successors$ is $K, L, M$:
+
+* Applies the interdiff between $base$ and $M$ to $L$ to produce $L'$
+* Applies the interdiff between $base$ and $L'$ to $K$ to produce $K'$
+* Returns $K'$
```

Comment:
I think we should keep it simple and limit divergence resolution to two commits at a time. Let's say the common predecessor is `A` and the two divergent commits are `B` and `C`. Since we have already asked the user to decide which parents to use (if there was a conflict in the first place), we can assume that `B` and `C` are siblings at this point. I think we should then first rebase `A` (just the tree, not the commit) to produce `A' = B⁻ + (A - A⁻)`  (where I used superscript minus for "parents of"). Then we produce the "resolved" content as `A' + (B - A') + (C - A') = B + (C - A')`.

Reply:
David and I talked about divergence resolution for quite some time yesterday. We think it should actually not make it harder to resolve divergence between all divergent commits at the same time. We would basically have a structure like this:

```
struct MergedState {
  parents: Merge<Vec<CommitId>>,
  author: Merge<Signature>,
  description: Merge<String>,
  tree: Merge<MergedTree>,
}
```

We initialize that based on the common predecessor. Then we apply the diff from each edge in the evolution graph.

Trees may need an additional step at the end, after we have asked the user which parent(s) to use.